contrib/zenith/pagestore_prefetch.c

```diff
@@ -0,0 +1,401 @@
+/*-------------------------------------------------------------------------
+ *
+ * pagestore_prefetch.c
+ *
+ *
+ * Portions Copyright (c) 1996-2021, PostgreSQL Global Development Group
+ * Portions Copyright (c) 1994, Regents of the University of California
+ *
+ *
+ * IDENTIFICATION
+ *	  contrib/zenith/pagestore_prefetch.c
+ *
+ *-------------------------------------------------------------------------
+ */
+#include "postgres.h"
+#include "miscadmin.h"
+#include "pgstat.h"
+#include "pagestore_client.h"
+#include "postmaster/bgworker.h"
+#include "storage/relfilenode.h"
+#include "storage/buf_internals.h"
+#include "storage/latch.h"
+#include "storage/ipc.h"
+#include "storage/lwlock.h"
+#include "utils/dynahash.h"
+#include "utils/guc.h"
+#include "utils/timestamp.h"
+#include "pagestore_client.h"
+
+typedef enum {
+	QUEUED,
+	IN_PROGRESS,
+	COMPLETED,
+	CANCELED,
+} PrefetchState;
+
+typedef struct PrefetchEntry {
+	BufferTag     tag;
+	XLogRecPtr    lsn;
+	uint32        index;  /* index of entry i prefetch cyclic buffer */
+	PrefetchState state;
+	Latch*        wait_latch; /* latch to be signaled when block is fetched */
+} PrefetchEntry;
+
+typedef struct PrefetchControl {
+	size_t curr; /* position in cyclic buffer */
+	Latch* go_latch; /* latch to wakeup prefetcher */
```

Comment:
If we always use this as an argument to `XxxLatch`, does it make sense to add a `_latch` prefix to this field? Overall, we have:

* go_latch
* wait_latch
* MyLatch (in this case wakeup event receiver is obvious)

I'd like to see more helpful names in the first two cases, e.g. `idle_prefetcher` & `idle_backend`.